Clean/bill

Francois389 opened the pull request:
Clean billing package

Add EmailService with HogMail
Add logger file

Francois389 committed:
feat: add Bill repository and service

- Introduces BillRepository extending JpaRepository for data access
- Adds BillService class annotated with @Service

Francois389 committed:
refactor: migrate date handling to LocalDate and JPA repository

- Replace Date with LocalDate for cleaner handling of dates
- Transition from BillDao and BillDaoImpl to BillRepository using Spring Data JPA
- Remove unused BillingService in favor of renamed BillService
- Update related tests to reflect infrastructure changes

Francois389 committed:
refactor: rename Billing to Bill across the project

Renames BillingController to BillController for consistency
Updates related test classes and method references

Ensures alignment with naming conventions used in the codebase
Improves readability and maintainability

Francois389 committed:
fix: remove getInstance

Francois389 committed:
feat: add SuccessfullResponseModel

Introduces a model for handling success responses
Includes message and success status attributes
Annotated for API documentation with Swagger

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Francois389 committed:
feat: enhance billing response and add exception handling

- Improve processBill and updatePrice to return SuccessfullResponseModel
- Replace direct string responses with structured response models
- Implement global exception handler for generic exceptions
- Refactor methods to support new response structures
- Extend integration tests for new response validation

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Francois389 committed:
feat: Update Bill and expand functionality

- Refactor Bill to use an enum for status, enhancing type safety
- Introduce BillService to encapsulate business logic and decouple from BillController
- Modify BillController to utilize BillService for fetching pending bills
- Enhance testing to validate new pending bill logic and ensure correct service integration

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Francois389 committed:
feat: add MedicalAct entity with price update logic

- Introduces MedicalAct entity to represent medical procedures
- Implements price update logic while marking old versions inactive
- Establishes basic unit tests for verifying price update functionality

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Francois389 committed:
feat: update BillDetail entity with MedicalAct integration

- Replaces treatmentName and unitPrice fields with MedicalAct association
- Streamlines lineTotal calculation based on MedicalAct pricing
- Updates tests to reflect changes in BillDetail structure

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Francois389 committed:
feat: add medical act data insertion

- Inserts new medical acts into the database
- Includes entries for consultation, x-ray, and surgery

Addresses data initialization for medical act table

Francois389 committed:
feat: add MedicalActService with repository and tests

- Introduces MedicalActService to manage medical acts
- Implements findByIds method in MedicalActService with validation
- Adds MedicalActRepository interface for data access
- Includes unit tests for service methods using Mockito
- Updates CustomAssertion with assertIsEmpty method

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Francois389 committed:
WIP feat: enhance billing process with medical acts

WIP

Mateofsr committed:
chore: move medical endpoint to new controller

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
feat: Endpoint for created MedicalAct

- with tests

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
fix: Remove query

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
fix: Use enum in parameter

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
fix: Tests

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
feat: Compute total revenue instead of field variable

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
fix: total revenue

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
chore: test for BillController Exception

To improve the coverage

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
chore: Extract method for bill creation

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
feat: remove filewriter for logger

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Mateofsr committed:
shame: fix after rebase

Co-Authored-By: François SP <75416431+Francois389@users.noreply.github.com>

Francois389 committed:
chore: update EmailService to use properties

- add HogMail for dev
- add GreenMail for test

Co-authored-by: Mateofsr <mateo.faussurier@iut-rodez.fr>

Odonata971 commented on src/main/java/sae/semestre/six/appointment/SchedulingController.java:
```diff
@@ -38,8 +38,9 @@ public SchedulingController(AppointmentRepository appointmentRepository, DoctorR
-    private final EmailService emailService = EmailService.getInstance();
+
+    @Autowired
+    private EmailService emailService;
```
use constructor injection insterad of variable injection

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillController.java:
```diff
@@ -0,0 +1,128 @@
+        this.emailService = emailService;
+    }
+
+    private BillController() {
```
delete empty useless constructor or put a comment to tell other devs that it is there to protect the public instantiation

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillController.java:
```diff
@@ -0,0 +1,128 @@
+            @Parameter(description = "List of medical acts id") @RequestParam Long[] medicalActId
+            ) throws Exception {
+
+            Patient patient = patientRepository.findById(Long.parseLong(patientId)).orElseThrow(
```
it is required to use architecture layers as Controller Service Repository. There should a call to a BillService method that handles the logic and do not call directly repo from controller.

Francois389 committed:
style: refactor code

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillController.java:
```diff
@@ -0,0 +1,128 @@
+    public InsuranceCoverageResponse calculateInsurance(
+            @Parameter(description = "Amount to cover") @RequestParam double amount
+            ) {
+        double coverage = amount;
```
what is the purpose of adding another variable that simply takes the amount given in parameter

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillController.java:
```diff
@@ -0,0 +1,128 @@
+    @GetMapping("/revenue")
+    public RevenueResponse getTotalRevenue() {
+        Double totalRevenue = billService.getTotalRevenue();
+        if (totalRevenue == null) {
```
same there, logic should be in service, even the if.

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillDetail.java:
```diff
@@ -1,58 +1,85 @@
-    
-    
+
+    public BillDetail() {
```
why adding it? It is created by default

Francois389 replied:
Not if we overrid it with another one : 
```java
    public BillDetail(Bill bill, MedicalAct medicalAct, Integer quantity) {
        this.bill = bill;
        this.medicalAct = medicalAct;
        this.quantity = quantity;
        calculateLineTotal();
    }
```

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillDetailService.java:
```diff
@@ -0,0 +1,27 @@
+public class BillDetailService {
+    private final BillDetailRepository billDetailRepository;
+
+    public BillDetailService(BillDetailRepository billDetailRepository) {
```
where is the @Autowired

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillDetailService.java:
```diff
@@ -0,0 +1,27 @@
+        this.billDetailRepository = billDetailRepository;
+    }
+
+    public List<BillDetail> updateBillDetail(MedicalAct old, MedicalAct newMedicalAct) {
```
Javadoc?

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillRepository.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDate;
+import java.util.List;
+
+interface BillRepository extends JpaRepository<Bill, Long> {
```
missing @Repository

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillService.java:
```diff
@@ -0,0 +1,61 @@
+    private final BillRepository billRepository;
+
+
+    public List<Bill> findPendingBills() {
```
javadoc?

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillService.java:
```diff
@@ -0,0 +1,61 @@
+        return billRepository.findBillsByStatus(Bill.Status.PENDING);
+    }
+
+    public void processBill(String patientId, String source, String[] items) {
```
empty?

Francois389 replied:
Yes 🗿

Francois389 replied:
Legacy code don't know what it's does

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillService.java:
```diff
@@ -0,0 +1,61 @@
+        return billRepository.findTotalRevenue();
+    }
+
+    public Bill processBill(Patient patient, Doctor doctor, List<MedicalAct> medicalActs) throws Exception {
```
same there and in all the project, put documentation on methods and isn't it possible to create Bill into its own constructor like `Bill(BillNumber, Patient, Doctor)` ?

Odonata971 commented on src/main/java/sae/semestre/six/appointment/doctor/Doctor.java:
```diff
@@ -35,7 +36,7 @@ public class Doctor {
+    private Set<Appointment> appointments = new HashSet<>();
 
     
     public Doctor() {
```
why

Francois389 replied:
why not ?

Francois389 replied:
We need an initialization for the `addAppoitment ` method

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActController.java:
```diff
@@ -0,0 +1,51 @@
+        this.medicalActRepository = medicalActRepository;
+    }
+
+    @GetMapping("/")
```
you can only put `@GetMapping` instead of adding `("/")`

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActController.java:
```diff
@@ -0,0 +1,51 @@
+        return new MedicalActResponse(medicalActRepository.save(medicalAct));
+    }
+
+    public record PricesResponse(List<MedicalAct> medicalActList) {
```
can't it be in a separate file?

Francois389 replied:
No

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.List;
+
+public interface MedicalActRepository extends JpaRepository<MedicalAct, Long> {
```
@Repository missing

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActRequest.java:
```diff
@@ -0,0 +1,16 @@
+package sae.semestre.six.appointment.medicalact;
```
Verify on all files that you added 
/*
 * FileName.java                                 dd month year
 * IUT de Rodez, no author rights
 */

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActService.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(MedicalActService.class);
+
+    public MedicalActService(MedicalActRepository medicalActRepository, BillService billService, BillDetailService billDetailService) {
```
where is the @Autowired for the constructor injection?

Francois389 committed:
fix: test that fail because of smtp server not online

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActService.java:
```diff
@@ -0,0 +1,74 @@
+        return medicalActRepository.findAllByActive(true);
+    }
+
+    public void updatePrice(double price, MedicalAct medicalAct) {
```
PLEASSSSSEEEE PUT JAVADOC

Odonata971 commented on src/main/java/sae/semestre/six/appointment/medicalact/MedicalActService.java:
```diff
@@ -0,0 +1,74 @@
+        // recalculate Bill
+    }
+
+    //private void recalculateAllPendingBills() throws Exception {
```
Its well known that you let code commented and push it to develop ...

Francois389 replied:
aaaaaaa

Odonata971 commented on src/main/java/sae/semestre/six/appointment/prescription/PrescriptionController.java:
```diff
@@ -37,6 +34,9 @@ public class PrescriptionController {
     private static final String AUDIT_FILE = "C:\\hospital\\prescriptions.log";
     
+    @Autowired
+    private BillService billService;
```
Use Constructor injection instead of field injection

Odonata971 commented on src/main/java/sae/semestre/six/appointment/prescription/PrescriptionController.java:
```diff
@@ -74,16 +74,16 @@ public String addPrescription(
-                .append(LocalDate.now().toString() + " - " + prescriptionId + "\n")
+                .append(LocalDate.now() + " - " + prescriptionId + "\n")
                 .close();
             
```
Not possible to have logic in a controller, put in a service or in the entity

Francois389 replied:
This file isn't in our scope

Odonata971 commented on src/main/java/sae/semestre/six/processor/MedicalBillingProcessor.java:
```diff
@@ -11,8 +11,8 @@ public class MedicalBillingProcessor {
+    private final List<String> pendingBills = new ArrayList<>();
     private double totalRevenue = 0.0;
 
     private MedicalBillingProcessor() {
```
add javadoc in the class

Francois389 replied:
This file isn't in our scope

Odonata971 commented on src/main/java/sae/semestre/six/stock/InventoryController.java:
```diff
@@ -21,8 +21,9 @@ public class InventoryController {
     private InventoryRepository inventoryRepository;
-    
-    private final EmailService emailService = EmailService.getInstance();
+
```
Use constructor injection of field injection

Odonata971 commented on src/main/resources/application.properties:
```diff
@@ -19,4 +19,23 @@ spring.jpa.show-sql=true
+# Mail
+spring.mail.host=localhost
+spring.mail.port=1025
+spring.mail.username=
```
add environment variables, even if it is in dev only for now, it will be in production

Odonata971 commented on src/test/java/sae/semestre/six/CustomAssertion.java:
```diff
@@ -47,4 +47,10 @@ public static void assertListNotEquals(List<?> expected, List<?> actual, String
         }
     }
+
+    public static void assertIsEmpty(List<?> list) {
```
J A V A D O C S      P L E A S E

Odonata971 commented on src/test/java/sae/semestre/six/appointment/bill/BillControllerIntegrationTest.java:
```diff
@@ -0,0 +1,288 @@
+    private AppointmentRepository appointmentRepository;
+
+    @Autowired
+    private EmailService emailService;
```
Use constructor injection of field injection

Francois389 replied:
No in test it's différent

Odonata971 commented on src/test/java/sae/semestre/six/appointment/bill/BillControllerIntegrationTest.java:
```diff
@@ -0,0 +1,288 @@
+    private EmailService emailService;
+
+    @Test
+    void testProcessBill() throws Exception {
```
add given when then on tests

Mateofsr replied:
tu te met un doigt dans l'oeil

Mateofsr replied:
jusqu'au coude si tu penses que je vais faire ton commentaire

Mateofsr replied:
bon je me suis calmé c'est pas dans mes habitudes d'être violent

Odonata971 commented on src/test/java/sae/semestre/six/appointment/bill/BillDetailServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Test
+    void testUpdateBillDetail() {
+        // Arrange
```
Use given when then

Odonata971 commented on src/test/java/sae/semestre/six/appointment/bill/BillRepositoryTest.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Test
+    void test() {
```
use given when then

Mateofsr replied:
then when given use

Odonata971 commented on src/test/java/sae/semestre/six/appointment/medicalact/MedicalActControllerIntegrationTest.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Test
+    void testUpdatePrice() throws Exception {
```
use given when then

Mateofsr replied:
like we say in french "rebelotte"

Odonata971 commented on src/test/java/sae/semestre/six/appointment/medicalact/MedicalActServiceTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Test
+    void findByIds_ShouldReturnMatchingMedicalActs_WhenIdsAreValid() {
+        // Arrange
```
use given when then

Mateofsr replied:
one more time

Odonata971 commented on src/test/java/sae/semestre/six/appointment/medicalact/MedicalActTest.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Test
+    void updatePrice() {
+        MedicalAct medicalAct = new MedicalAct("Test Act", 100.0);
```
use given when then

Mateofsr replied:
ratio

Odonata971 reviewed (changes requested):
Don't forget javadoc everywhere AND.... and... use given when then along with constructor injection instead of field injection

Francois389 committed:
feat: enhance billing logic and medical act integration

- Adds Bill.recalculate method to update the total for pending bills
- Calculates line total when setting a medical act in BillDetail
- Modifies BillRepository interface visibility to public
- Integrates recalculate logic in MedicalActService to update affected bills after price change
- Introduces test for updating price with bill recalculation

Francois389 committed:
chore: update log file path

Moves billing log file to the 'logs' directory for better organization
and to adhere to directory structure standards.

Francois389 committed:
fix: review

Francois389 committed:
fix: typo

Odonata971 commented on src/main/java/sae/semestre/six/appointment/Appointment.java:
```diff
@@ -46,11 +46,11 @@ public class Appointment {
 
-    
+
     public Appointment() {
```
delete that

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/Bill.java:
```diff
@@ -14,70 +14,152 @@
+        return billDetails;
+    }
+
+    public Bill addBillDetail(BillDetail billDetail) {
```
add javadoc

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/Bill.java:
```diff
@@ -14,70 +14,152 @@
+        return this;
+    }
+
+    public enum Status {
```
can't it be a separate file ?

Francois389 replied:
I don't think so, it's only been use in the Bill class.

Francois389 committed:
fix: remove default constructor declaration

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillController.java:
```diff
@@ -0,0 +1,121 @@
+package sae.semestre.six.appointment.bill;
```
add the package and rights

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillDetail.java:
```diff
@@ -1,58 +1,86 @@
 package sae.semestre.six.appointment.bill;
```
missing package and rights

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillDetailService.java:
```diff
@@ -0,0 +1,36 @@
+package sae.semestre.six.appointment.bill;
```
missing package and rights

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillRepository.java:
```diff
@@ -0,0 +1,26 @@
+package sae.semestre.six.appointment.bill;
```
missing package and rights

Odonata971 commented on src/main/java/sae/semestre/six/appointment/bill/BillService.java:
```diff
@@ -0,0 +1,76 @@
+package sae.semestre.six.appointment.bill;
```
missing package and rights

Odonata971 commented on src/main/java/sae/semestre/six/appointment/patient/history/PatientHistory.java:
```diff
@@ -129,27 +127,27 @@ public void setNotes(String notes) {
+        this.appointments = appointments;
+    }
+
     public List<Bill> getBillsSorted() {
```
javadoc

Francois389 replied:
trivial

Odonata971 commented on src/main/java/sae/semestre/six/generic/AbstractHibernateDao.java:
```diff
@@ -8,22 +8,22 @@
 import java.util.List;
 
 public abstract class AbstractHibernateDao<T, ID extends Serializable> implements GenericDao<T, ID> {
```
delete file

Odonata971 commented on src/main/java/sae/semestre/six/generic/DaoGenerique.java:
delete file

Odonata971 commented on src/main/java/sae/semestre/six/generic/DaoHibernateAbstrait.java:
delete file

Odonata971 commented on src/main/java/sae/semestre/six/generic/GenericDao.java:
delete file

Odonata971 reviewed (commented):
changes these and it's all good

Francois389 committed:
feat: add header

Francois389 committed:
chore: remove Generic

+ fix test

Francois389 committed:
fix: add missing slash in url

Francois389 committed:
fix: remove vcs.xml

Francois389 committed:
feat: add JavaDoc

Francois389 committed:
fix: rename class declaration

Francois389 committed:
fix: endpoint

Francois389 committed:
fix: sonarcloud issue

Odonata971 reviewed (dismissed):
👍🏼

Francois389 committed:
feat: add todo

Odonata971 reviewed (approved):
👍🏼 👍
